src/app-layer-smtp.c

```diff
@@ -1277,43 +1278,44 @@ static int SMTPProcessRequest(SMTPState *state, Flow *f,
 
 static int SMTPPreProcessCommands(SMTPState *state, Flow *f, AppLayerParserState *pstate)
 {
-    // By this time we should have had the command line parsed
-    uint8_t *lf_idx = memchr(state->input + state->consumed, 0x0a, state->input_len);
-    const uint32_t orig_input_len = state->input_len;
-    // Both DATA and BDAT set SMTP_PARSER_STATE_COMMAND_DATA_MODE, so this while
-    // loop should be valid for both
-    while (state->input_len > 0 && (state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE)) {
-        uint8_t delim_len = 0;
-        uint32_t consumed_line = 0;
-        if (lf_idx == NULL) {
-            if ((state->input_len == 1 && state->input[state->consumed] == '-') ||
-                    (state->input_len > 1 && state->input[state->consumed] == '-' &&
+    bool line_complete = false;
+    int32_t input_len = state->input_len;
+    for (int32_t i = 0;
+            i < input_len && (state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE); i++) {
```

Comment:
can this be unset in this loop? Perhaps in the `SMTPProcessRequest` call? If so, lets just check for it after that call then, and break the loop?

Reply:
thinking about
```diff
diff --git a/src/app-layer-smtp.c b/src/app-layer-smtp.c
index fe441c778..c19398ad8 100644
--- a/src/app-layer-smtp.c
+++ b/src/app-layer-smtp.c
@@ -1278,10 +1278,11 @@ static int SMTPProcessRequest(SMTPState *state, Flow *f,
 
 static int SMTPPreProcessCommands(SMTPState *state, Flow *f, AppLayerParserState *pstate)
 {
+    DEBUG_VALIDATE_BUG_ON((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0);
+
     bool line_complete = false;
     int32_t input_len = state->input_len;
-    for (int32_t i = 0;
-            i < input_len && (state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE); i++) {
+    for (int32_t i = 0; i < input_len; i++) {
         if (state->input[i] == 0x0d) {
             if (i < input_len - 1 && state->input[i + 1] == 0x0a) {
                 i++;
@@ -1315,6 +1316,9 @@ static int SMTPPreProcessCommands(SMTPState *state, Flow *f, AppLayerParserState
             }
             line_complete = false;
             state->current_line_delimiter_len = 0;
+
+            if ((state->parser_state & SMTP_PARSER_STATE_COMMAND_DATA_MODE) == 0)
+                break;
         }
     }
     return 0;
```
which I find more readable and easier to understand